src/lib.rs

```diff
@@ -193,197 +195,212 @@ where
+      None,
+    );
+    let circuit_secondary: NovaAugmentedCircuit<G1, C2> = NovaAugmentedCircuit::new(
+      pp.augmented_circuit_params_secondary.clone(),
+      Some(inputs_secondary),
+      c_secondary.clone(),
+      pp.ro_consts_circuit_secondary.clone(),
+    );
+    let _ = circuit_secondary.synthesize(&mut cs_secondary);
+    let (u_secondary, w_secondary) = cs_secondary
+      .r1cs_instance_and_witness(&pp.r1cs_shape_secondary, &pp.ck_secondary)
+      .map_err(|_e| NovaError::UnSat)
+      .expect("Nova error unsat");
+
+    // IVC proof for the primary circuit
+    let l_w_primary = w_primary;
+    let l_u_primary = u_primary;
+    let r_W_primary = RelaxedR1CSWitness::from_r1cs_witness(&pp.r1cs_shape_primary, &l_w_primary);
+    let r_U_primary =
+      RelaxedR1CSInstance::from_r1cs_instance(&pp.ck_primary, &pp.r1cs_shape_primary, &l_u_primary);
+
+    // IVC proof of the secondary circuit
+    let l_w_secondary = w_secondary;
+    let l_u_secondary = u_secondary;
+    let r_W_secondary = RelaxedR1CSWitness::<G2>::default(&pp.r1cs_shape_secondary);
+    let r_U_secondary =
+      RelaxedR1CSInstance::<G2>::default(&pp.ck_secondary, &pp.r1cs_shape_secondary);
+
+    // Outputs of the two circuits thus far
+    let zi_primary = c_primary.output(&z0_primary);
+    let zi_secondary = c_secondary.output(&z0_secondary);
+
+    if zi_primary.len() != pp.F_arity_primary || zi_secondary.len() != pp.F_arity_secondary {
+      panic!("Invalid step length");
+    }
+
+    Self {
+      r_W_primary,
+      r_U_primary,
+      l_w_primary,
+      l_u_primary,
+      r_W_secondary,
+      r_U_secondary,
+      l_w_secondary,
+      l_u_secondary,
+      i: 0,
+      zi_primary,
+      zi_secondary,
+      _p_c1: Default::default(),
+      _p_c2: Default::default(),
+    }
+  }
+
   /// Create a new `RecursiveSNARK` (or updates the provided `RecursiveSNARK`)
   /// by executing a step of the incremental computation
   pub fn prove_step(
+    &mut self,
     pp: &PublicParams<G1, G2, C1, C2>,
-    recursive_snark: Option<Self>,
-    c_primary: C1,
-    c_secondary: C2,
+    c_primary: &C1,
+    c_secondary: &C2,
     z0_primary: Vec<G1::Scalar>,
     z0_secondary: Vec<G2::Scalar>,
-  ) -> Result<Self, NovaError> {
+  ) -> Result<(), NovaError> {
     if z0_primary.len() != pp.F_arity_primary || z0_secondary.len() != pp.F_arity_secondary {
       return Err(NovaError::InvalidInitialInputLength);
     }
 
-    match recursive_snark {
-      None => {
-        // base case for the primary
-        let mut cs_primary: SatisfyingAssignment<G1> = SatisfyingAssignment::new();
-        let inputs_primary: NovaAugmentedCircuitInputs<G2> = NovaAugmentedCircuitInputs::new(
-          scalar_as_base::<G1>(pp.digest),
-          G1::Scalar::ZERO,
-          z0_primary.clone(),
-          None,
-          None,
-          None,
-          None,
-        );
+    // Frist step was already done in the constructor
+    if self.i == 0 {
```

Comment:
Does it make sense to set i to 1 in the constructor?

Reply:
I tried, the problem is wen I do this we might need to change the entire code to staring prove from step 1.

```rust
for i in 1..step_numbers {
  recursive_snark.prove_step(...);
}
```

Reply:
Good point! Let me look at the code again and see if there is a way to fix this without having to change the step numbers.

Reply:
@chiro-hiro @huitseeker After thinking more, I think changing prove_step to only prove from 1..step_numbers might be the cleanest option. Could we do this?

Also, we just merged a PR from @huitseeker to address clippy issues. Can we please rebase this branch?

Reply:
I've put up a commit to sketch what that change would look like (don't use as-is, it contains a couple of commented lines that should be cleaned up):
https://github.com/huitseeker/Nova/commit/6342a263b80a347a4e3a252b88bc8dcb4037eca7

I'm not convinced that changing proof iterations to run [1..n] is more ergonomic than the idempotent / backwards-compatible approach of the present PR, which makes miscounting steps hard. But YMMV - and I hope the commit is directionally useful if @srinathsetty thinks this change should happen.

Reply:
I'm prefer to start from `0` instead of `1`. If we want to solve this completely we might need to change a lot of code , and change the `RecursiveSNARK`'s context to optional `Option<RecursiveSNARKContext>` to allow us init a new instance without prove anything.

Reply:
Thanks, @huitseeker and @chiro-hiro! We can keep the idempotent approach in the present PR.

Reply:
> If we want to solve this completely we might need to change a lot of code , and change the RecursiveSNARK's context to optional Option<RecursiveSNARKContext> to allow us init a new instance without prove anything.

I think it's an interesting idea, that may be worth opening an issue @chiro-hiro : it's also something we can perhaps engineer with a `OnceCell`, especially since they're now stabilized in Rust 1.70.